channel_store.go

```diff
@@ -0,0 +1,126 @@
+package sockchat
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+)
+
+var ErrChannelDoesNotExist = errors.New("channel does not exist")
+var ErrUserNotInChannel = errors.New("user is not member of this channel")
+var ErrUserAlreadyInChannel = errors.New("user is already member of this channel")
+var ErrEmptyChannelName = errors.New("channel's `name` is missing")
+
+type ChannelStore struct {
+	Channels map[string]*Channel
+	lock     sync.RWMutex
+}
+
+func NewChannelStore() (*ChannelStore, error) {
+	return &ChannelStore{Channels: make(map[string]*Channel)}, nil
+}
+
+func (s *ChannelStore) GetChannel(name string) (*Channel, error) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+	channel := s.Channels[name]
+	if channel == nil {
+		return nil, ErrChannelDoesNotExist
+	}
+	return channel, nil
+}
+
+func (s *ChannelStore) CreateChannel(channelName string) error {
+	if channelName == "" {
+		return ErrEmptyChannelName
+	}
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	if s.Channels[channelName] != nil {
+		return fmt.Errorf("channel `%s` already exists", channelName)
+	}
+	s.Channels[channelName] = &Channel{members: make(map[SockchatUserHandler]bool)}
+	return nil
+}
+
+func (s *ChannelStore) AddUserToChannel(channelName string, user SockchatUserHandler) error {
+	if channelName == "" {
+		return ErrEmptyChannelName
+	}
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	channel := s.Channels[channelName]
+	if channel == nil {
+		return ErrChannelDoesNotExist
+	}
+	if channel.members[user] {
+		return ErrUserAlreadyInChannel
+	}
+	channel.AddMember(user)
+	channel.MessageMembers(NewSocketMessage(UserJoinedChannelEvent, ChannelUserChangeEvent{channelName, user.getNick()}))
+	return nil
+}
+
+func (s *ChannelStore) RemoveUserFromChannel(channelName string, user SockchatUserHandler) error {
+	if channelName == "" {
+		return ErrEmptyChannelName
+	}
+	channel, err := s.GetChannel(channelName)
+	if err != nil {
+		return err
+	}
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	if !channel.members[user] {
+		return ErrUserNotInChannel
+	}
+	delete(channel.members, user)
```

Comment:
```suggestion
	channel.RemoveMember(user)
```
This delete operation without lock should be fixed

Reply:
To more precise: the lock in this function has been on the whole ChannelStore, but not on the Channel, which could be accessed by `MessageMembers` in other goroutines